### PR TITLE
DAOS-13214 control: Disallow dRPCs if engine is not ready

### PR DESCRIPTION
**Pull request by tanabarr**
Move IsReady() check into CallDrpc() function and adjust rank setup
routine as appropriate. This eliminates the possibility of calling
into an engine from the control-plane before SetRank() and Setup()
have both returned successfully.

Required-githooks: true

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by kjacque — DISMISSED** [review]

File: src/control/server/instance.go [hunk 1]
```diff
@@ -287,7 +287,7 @@ func (ei *EngineInstance) SetupRank(ctx context.Context, rank ranklist.Rank, map
 }
 
 func (ei *EngineInstance) callSetRank(ctx context.Context, rank ranklist.Rank, map_version uint32) error {
-	dresp, err := ei.CallDrpc(ctx, drpc.MethodSetRank, &mgmtpb.SetRankReq{Rank: rank.Uint32(), MapVersion: map_version})
+	dresp, err := ei.callDrpc(ctx, drpc.MethodSetRank, &mgmtpb.SetRankReq{Rank: rank.Uint32(), MapVersion: map_version})
```
Comment (liw) [inline src/control/server/instance.go:290]: Ah, this PR supersedes the previous one? I think this is a good solution too.

**Review by liw — DISMISSED** [review]

**Review by knard-intel — DISMISSED** [review]

File: src/control/server/harness.go [hunk 2]
```diff
@@ -186,22 +186,10 @@ func (h *EngineHarness) CallDrpc(ctx context.Context, method drpc.Method, body p
...
-					err = FaultDataPlaneNotStarted
-				}
-			}
-			continue
-		}
 		resp, err = i.CallDrpc(ctx, method, body)
 
 		switch errors.Cause(err) {
-		case errDRPCNotReady, FaultDataPlaneNotStarted:
+		case errEngineNotReady, errDRPCNotReady:
```
Comment (mjmac) [inline src/control/server/harness.go:192]: I think that this subtle change is what's causing the MS Failover test problem. The logic in this method was intended to invoke the dRPC failure handlers when the MS Leader has no ranks available to service a request. Looking at the control logs, we can see that the MS Leader's rank is killed, but it never resigns leadership, so the PoolCreate RPC is just retried over and over until timeout. What's supposed to happen is that the PoolCreate goes to the current MS Leader after its rank is killed, the RPC fails and the dRPC failure handlers are invoked (including a ResignLeadership() call), and the PoolCreate is then serviced by the new MS Leader.

I'm not sure what the best approach here is -- basically you're going to want to distinguish between the following scenarios:

- Engine wasn't started yet (retryable)
- Engine was started, but there was a race and the dRPC connection wasn't ready yet (retryable)
- Engine was started, but the dRPC call failed (not retryable, return the error to the caller)

I suspect that the changes in this PR have eliminated detection of the last scenario, with the result that the MS Leader never fails over.

Comment (tanabarr) [reply src/control/server/harness.go:192]: Thanks Mike, let me try to fix

Comment (tanabarr) [reply src/control/server/harness.go]: So should harness.CallDrpc() specifically avoid checking if the instances are ready before calling instance.callDrpc()?

**Review by mjmac — APPROVED** [review]

**Review by kjacque — APPROVED** [review]

File: src/control/server/harness.go (hunk 2 above)
Comment (tanabarr) [reply src/control/server/harness.go]: I've updated implementation

**Review by liw — APPROVED** [review]

**Review by knard-intel — APPROVED** [review]

**Pull request merged**